monai/metrics/cumulative_average.py

```diff
@@ -9,64 +9,168 @@
-    It supports single class or multi-class data, for example,
-    value can be 0.44 (a loss value) or [0.3, 0.4] (metrics of two classes).
-    It also supports distributed data parallel, sync data when aggregating.
-    For example, recording loss values and compute the overall average value in every 5 iterations:
+    A utility class to keep track of average values. For example during training/validation loop,
+    we need to accumulate the per-batch metrics and calculate the final average value for the whole dataset.
+    When training in multi-gpu environment, with DistributedDataParallel, it will average across the processes.
+
+    Example:
 
     .. code-block:: python
 
-        average = CumulativeAverage()
-        for i, d in enumerate(dataloader):
-            loss = ...
-            average.append(loss)
-            if i % 5 == 0:
-                print(f"cumulative average of loss: {average.aggregate()}")
-        average.reset()
+        from monai.metrics import CumulativeAverage
+
+        run_avg = CumulativeAverage()
+        for i in range(len(train_set)):
+            ...
+            val = calc_metric(x,y) #some metric value
+            run_avg.append(val)
+
+        val_avg = run_avg.aggregate() #average value
 
     """
 
-    def __init__(self) -> None:
-        super().__init__()
-        self.sum = None
-        self.not_nans = None
+    def __init__(self, group: Optional[ProcessGroup] = None) -> None:
+        """
+        Args:
+            group: Optional process group to use in multi-gpu distributed setup. Defaults to None
+            to use the default process group created during DistributedDataParallel initialization.
+        """
 
-    def reset(self):
+        self.is_distributed = dist.is_available() and dist.is_initialized()
+        self.device = None
+
+        if self.is_distributed:
+            self.group = group
+            if dist.get_backend(group=group) == dist.Backend.NCCL:
+                self.device = torch.device("cuda", dist.get_rank(group=group))
+        else:
+            self.group = None
+
+        if self.device is None:
+            self.device = torch.device("cpu")
+
+        self.reset()
+
+    def reset(self) -> None:
         """
-        Reset all the running status, including buffers, sum, not nans count, etc.
+        Reset all the running status
 
         """
-        super().reset()
-        self.sum = None
-        self.not_nans = None
+        self.val: torch.Tensor = 0  # type: ignore
+        self.sum: torch.Tensor = 0  # type: ignore
+        self.count: torch.Tensor = 0  # type: ignore
 
-    def aggregate(self):
+    def proper_tensor(self, a: Any) -> torch.Tensor:
         """
-        Sync data from all the ranks and compute the average value with previous sum value.
+        Ensure torch.Tensor float format and optionally copy to the proper gpu device
+        """
+        x: torch.Tensor = a.detach() if isinstance(a, torch.Tensor) else torch.tensor(a)
```

Comment:
Normally the metrics module shouldn't change the device of the input if we follow pytorch's api style. The user should prepare the input beforehand accordingly in a compatible shape/type/device... If this is 100% needed, we may have to change all the relevant APIs just to me consistent

Reply:
this line is just to ensure the the avergage value is on the correct device ( CPU vs GPU).  If a user provide it as a simple constant or a list (on cpu and we're in DDP on gpu), it'll be converted to proper GPU device. And otherwise if we're not in DDP, it will ensure we're maintaining average on CPU.  

many pytorch api accepts input as either a constant or a tensor, and support shape broadcasting. In either case this  self.proper_tensor()  is just an internal sub-function.

Reply:
sorry, the comment is meant to be on the the `self.device` usage in general in this class

Reply:
@wyli thank you for your comment. the self.device is set according to the processing group. If the user specified group on CPU, then device is on CPU, if the user specified it on GPU (or it's default group on GPU/ NCCL) then the device is GPU.  We need to insure the proper device to be able to run multiprocessing reduction on proper Processing group in DDP.  Please let me know if there is something specific you'd like me to change in the code.  thank you

Reply:
@wyli , should I mark this conversation as resolved? thank you

Reply:
Thanks, I don't think this class should assign the devices, it doesn't even consider the difference between rank vs local rank.

Reply:
@wyli alright, I've removed self.device , it will be up to a user now to use the correct device.  Let me know if you have any other objections.